src/napari/_vispy/visuals/scale_bar.py

```diff
@@ -1,76 +1,72 @@
 import numpy as np
 from vispy.scene.visuals import Compound, Line, Rectangle, Text
 
+from napari._vispy.utils.text import get_text_font_size, get_text_width_height
+
 
 class ScaleBar(Compound):
     def __init__(self) -> None:
-        self._data = np.array(
+        self._line_data = np.array(
             [
-                [0, 0],
+                [-1, 0],
                 [1, 0],
-                [0, -5],
-                [0, 5],
-                [1, -5],
-                [1, 5],
+                [-1, -1],
+                [-1, 1],
+                [1, -1],
+                [1, 1],
             ]
         )
 
-        # order matters (last is drawn on top)
-        super().__init__(
-            [
-                Rectangle(center=[0.5, 0.5], width=1.1, height=36),
-                Text(
-                    text='1px',
-                    pos=[0.5, 0.5],
-                    anchor_x='center',
-                    anchor_y='top',
-                    font_size=10,
-                ),
-                Line(connect='segments', method='gl', width=3),
-            ]
-        )
-
-    @property
-    def line(self):
-        return self._subvisuals[2]
+        self._box_padding = 6
+        self._tick_length = 11  # odd numbers look better
+        self._color = (1, 1, 1, 1)
+        self._box_color = (0, 0, 0, 1)
+        self._text_vertices_size = (0, 0)
 
-    @property
-    def text(self):
-        return self._subvisuals[1]
+        self.box = Rectangle(center=[0.5, 0.5], width=100, height=36)
+        self.text = Text(
+            text='1px',
+            pos=[0.5, 0.5],
+            anchor_x='center',
+            anchor_y='bottom',
+            font_size=10,
+        )
+        self.line = Line(connect='segments', method='gl', width=3)
+        # order matters (last is drawn on top)
+        super().__init__([self.box, self.text, self.line])
 
-    @property
-    def box(self):
-        return self._subvisuals[0]
+    def set_data(self, *, length, color, ticks, font_size):
+        text_width, _ = get_text_width_height(self.text)
```

Comment:
```suggestion
        text_width, _ = 10, 0
```
I do not understand why, but this line is causing the reverse order of layers.